aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -2175,6 +2187,56 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
   }
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ linalg_eigh ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+// This is a type dispatch function for 'apply_magma_eigh'
+void linalg_eigh_magma(Tensor& eigenvalues, Tensor& eigenvectors, Tensor& infos, bool upper, bool compute_eigenvectors) {
+  // MAGMA just calls LAPACK for n <= 128
+  // See https://bitbucket.org/icl/magma/src/e6fdca447bd402693e8b0b950a898b6879bbcc41/src/zheevd_gpu.cpp?at=master#lines-258
+  // in addition it does bad stuff ignoring lda breaking 0x0 inputs
+  bool use_magma = true;
+  if (eigenvectors.size(-1) <= 128) {
+    use_magma = false;
+  }
+
+  if (use_magma) {
+    // MAGMA requires eigenvalues and infos tensors to reside on CPU
+    Tensor eigenvalues_cpu = eigenvalues.to(kCPU);
+    infos = infos.to(kCPU);
+
+    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(
+      eigenvectors.scalar_type(), "linalg_eigh_cpu", [&] {
+        apply_magma_eigh<scalar_t>(
+            eigenvalues_cpu, eigenvectors, infos, upper, compute_eigenvectors);
+      });
+
+    // Transfer computed by MAGMA results from CPU to GPU
+    eigenvalues.copy_(eigenvalues_cpu);
+  } else { 
+    // transfer to CPU, compute the result and copy back to GPU
+    // this is faster than going through MAGMA that does the same
+    Tensor eigenvalues_cpu = at::empty_like(eigenvalues, eigenvalues.options().device(kCPU));
+    if (compute_eigenvectors) {
+      Tensor eigenvectors_cpu = at::empty_like(eigenvectors, eigenvectors.options().device(kCPU));
+      at::linalg_eigh_out(eigenvalues_cpu, eigenvectors_cpu, eigenvectors.to(kCPU), upper ? "U" : "L");
+      eigenvectors.copy_(eigenvectors_cpu);
+    } else {
+      at::linalg_eigvalsh_out(eigenvalues_cpu, eigenvectors.to(kCPU), upper ? "U" : "L");
+    }
+    eigenvalues.copy_(eigenvalues_cpu);
+  }
+}
+
+void linalg_eigh_kernel(Tensor& eigenvalues, Tensor& eigenvectors, Tensor& infos, bool upper, bool compute_eigenvectors) {
+#if defined(USE_CUSOLVER)
+  linalg_eigh_cusolver(eigenvalues, eigenvectors, infos, upper, compute_eigenvectors);
```

Comment:
So if cuSOLVER is available then it will be exclusively be used, and the algorithm occurs entirely on the GPU?

Reply:
That's right.